pinot-core/src/test/java/org/apache/pinot/core/segment/processing/framework/SegmentProcessorFrameworkTest.java

```diff
@@ -581,6 +581,51 @@ public void testMultipleSegments()
     rewindRecordReaders(_multipleSegments);
   }
 
+  @Test
```

Comment:
Could you cover following scenarios
1. Default case where there's no limit
2. Multiple readers where there's limit
3. Single reader where there's limit
4. Delegate Record Readers Vs Regular record readers (created in RecordReaderFileConfig), to ensure that we handle delegate readers correctly. 
5. Case where there's limit and there's failure, to make sure that things are cleaned up (eg: intermediate mapper files) -> Not sure if we have test for this but will be good to verify, at least manually. 
6. Ensure that we allocate/free reader and there's no leak.

Reply:
Added all the tests for each of the following cases :

1. default case there is no limit (single recorReader and multiple recordReader)
2. Single record reader when there is a limit.
3. Multiple record readers when there is a limit.
4. Case where there's a limit and there is a failure to check if the files are cleaned up.
5. Delegate record readers vs regular record readers (check if we are closing in the case of regular ones and check if we are not closing in case of delegate record readers) for both with limit and without limit.